homeassistant/components/gardena/__init__.py

```diff
@@ -0,0 +1,99 @@
+"""Support for Gardena Smart system devices."""
+import logging
+
+from gardena.smart_system import SmartSystem
+from oauthlib.oauth2.rfc6749.errors import (
+    AccessDeniedError,
+    InvalidClientError,
+    MissingTokenError,
+)
+import voluptuous as vol
+
+from homeassistant.const import CONF_PASSWORD, CONF_EMAIL
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "gardena"
+GARDENA_SYSTEM = "gardena_system"
+GARDENA_LOCATION = "gardena_location"
+GARDENA_CONFIG = "gardena_config"
+CONF_CLIENT_ID = "client_id"
+CONF_LOCATION_ID = "location_id"
+CONF_MOWER_DURATION = "mower_duration"
+CONF_SMART_IRRIGATION_DURATION = "smart_irrigation_control_duration"
+CONF_SMART_WATERING = "smart_watering_duration"
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_EMAIL): cv.string,
+                vol.Required(CONF_PASSWORD): cv.string,
+                vol.Required(CONF_CLIENT_ID): cv.string,
+                vol.Required(CONF_LOCATION_ID): cv.string,
+                vol.Optional(CONF_MOWER_DURATION, default="60"): cv.string,
+                vol.Optional(CONF_SMART_IRRIGATION_DURATION, default="60"): cv.string,
+                vol.Optional(CONF_SMART_WATERING, default="60"): cv.string,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+ATTR_NAME = "name"
+ATTR_ACTIVITY = "activity"
+ATTR_BATTERY_STATE = "battery_state"
+ATTR_RF_LINK_LEVEL = "rf_link_level"
+ATTR_RF_LINK_STATE = "rf_link_state"
+ATTR_SERIAL = "serial"
+ATTR_OPERATING_HOURS = "operating_hours"
+ATTR_LAST_ERRORS = "last_error"
+
+
+async def async_setup(hass, config):
+    """Set up the Gardena integation."""
+    _LOGGER.debug("Initialising Gardena")
+
+    try:
+        hass.data[GARDENA_SYSTEM] = GardenaSmartSystem(
```

Comment:
If this can raise exceptions, it means that you're doing I/O inside the event loop. That is not allowed. Run in executor instead. `hass.data[GARDENA_SYSTEM] = await hass.async_add_executor_job(GardenaSmartSystem, hass, config[DOMAIN], SmartSystem)`